schema/pgsql/schema.sql

```diff
@@ -1,5 +1,14 @@
 CREATE TYPE boolenum AS ENUM ( 'n', 'y' );
-CREATE TYPE incident_history_event_type AS ENUM ( 'incident_severity_changed', 'recipient_role_changed', 'escalation_triggered', 'rule_matched', 'opened', 'closed', 'notified' );
+CREATE TYPE incident_history_event_type AS ENUM (
+    -- Order to be honored for events with identical microsecond timestamps.
+    'opened',
+    'recipient_role_changed',
+    'incident_severity_changed',
+    'rule_matched',
+    'escalation_triggered',
+    'notified',
+    'closed'
```

Comment:
Did you have a specific reason to put `recipient_role_changed` where it is? I'd say it should come after `escalation_triggered` as it can happen as part of an escalation triggering:

https://github.com/Icinga/icinga-notifications/blob/805f114d3f6f390a9299c3dd7ac47f08a0ce5bda/internal/incident/sync.go#L77-L110

Reply:
Valid point.

Furthermore, I tried to inspect `icinga-notification`'s current behavior with the following patch, inspected by the script below, resulting in multiple blocks like the posted on. <sup>(Yes, I overlooked the option to perform a SQL query based on `incident_history.id`)</sup>

Thus, I have also moved `incident_severity_changed` down and put `notified` last as `notified` might also happen after a `closed`.

```diff
diff --git a/internal/incident/sync.go b/internal/incident/sync.go
index b29969d..162cbbd 100644
--- a/internal/incident/sync.go
+++ b/internal/incident/sync.go
@@ -38,6 +38,10 @@ func (i *Incident) Sync(ctx context.Context, tx *sqlx.Tx) error {
 func (i *Incident) AddHistory(ctx context.Context, tx *sqlx.Tx, historyRow *HistoryRow, fetchId bool) (types.Int, error) {
        historyRow.IncidentID = i.incidentRowID

+       i.logger.Debugw("Adding incident history",
+               zap.Int64("incident_id", historyRow.IncidentID),
+               zap.Stringer("history_event_type", &historyRow.Type))
+
        stmt := utils.BuildInsertStmtWithout(i.db, historyRow, "id")
        if fetchId {
                historyId, err := utils.InsertAndFetchId(ctx, tx, stmt, historyRow)
```

```sh
$ docker logs notifications 2>&1 \
  | gawk '
      /Adding incident history/ {
        match($0, /incident_id": ([0-9]+),.*history_event_type": "([a-z_]+)"/, arr)
        print NR, arr[1], $1, arr[2]
      }' \
  | sort -n -k 2 -k 1 \
  | gawk '
      prev_incident_id != $2 {
        prev_incident_id = $2
        print "#" $2
      }
      { print "  - " $3 " " $4 }'
```

```
#1416
  - 2024-04-23T09:27:25.773Z opened
  - 2024-04-23T09:27:25.776Z rule_matched
  - 2024-04-23T09:27:25.779Z escalation_triggered
  - 2024-04-23T09:27:25.782Z notified
  - 2024-04-23T09:28:25.782Z escalation_triggered
  - 2024-04-23T09:28:25.784Z notified
  - 2024-04-23T09:31:14.805Z incident_severity_changed
  - 2024-04-23T09:31:14.806Z closed
  - 2024-04-23T09:31:14.807Z notified
  - 2024-04-23T09:31:14.807Z notified
```

Reply:
> Thus, I have also moved `incident_severity_changed` down

I don't see why that's the case from the output you shared. I had expected `incident_severity_changed` before `escalation_triggered` as you can trigger escalations based on the incident severity.

Reply:
It appeared at this place as for closing incidents an `incident_severity_changed`  `incident_history` entries appears straight before a `closed` one, both issued in [`Incident.processSeverityChangedEvent`](https://github.com/Icinga/icinga-notifications/blob/462bc6c960bb029f93be342aebb06eb630a54cbf/internal/incident/incident.go#L267).

In its calling method, the following order is honored:
1. `Incident.processSeverityChangedEvent`: `incident_severity_changed`.
2. `Incident.evaluteRules`: `rule_matched`
3. `Incident.triggerEscalations`: `escalation_triggered`

https://github.com/Icinga/icinga-notifications/blob/462bc6c960bb029f93be342aebb06eb630a54cbf/internal/incident/incident.go#L163-L185